# Personal Finance Analyzer


OBJECTIVE:
Analyze transaction data to uncover spending patterns, financial risks, and provide strategic recommendations.

PHASE 1 — DATA AUDIT & CLEANING

In [1]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
# 1. Load the dataset and inspect structure:
df = pd.read_csv("dataset.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Date                     1500 non-null   object 
 1   Transaction Description  1500 non-null   object 
 2   Category                 1500 non-null   object 
 3   Amount                   1500 non-null   float64
 4   Type                     1500 non-null   object 
dtypes: float64(1), object(4)
memory usage: 58.7+ KB


In [3]:
df.head(1), df.tail(1)

(         Date Transaction Description      Category   Amount     Type
 0  2020-01-02             Score each.  Food & Drink  1485.69  Expense,
             Date            Transaction Description Category  Amount    Type
 1499  2024-12-29  Pressure activity defense detail.    Other  1480.0  Income)

In [4]:
# 2. Data Quality Checks:
df.isna().values.any(), df.isnull().values.any()

(np.False_, np.False_)

In [5]:
df.duplicated().any()

np.False_

In [6]:
(df["Amount"] < 0).any()

np.False_

In [7]:
df.groupby("Type").sum()

,Date,Transaction Description,Category,Amount
Type,,,,
Expense,2020-01-022020-01-022020-01-042020-01-132020-0...,Score each.Quality throughout.Instead ahead de...,Food & DrinkUtilitiesRentFood & DrinkShoppingF...,1227194.37
Income,2020-01-052020-01-192020-02-052020-02-082020-0...,Information last everything thank serve.Court ...,InvestmentOtherInvestmentOtherOtherInvestmentI...,734087.00


In [8]:
# 3. Date Handling:
df["Date"] = pd.to_datetime(df["Date"])

In [9]:
df["Date"].isna().values.any(), df["Date"].isnull().values.any()

(np.False_, np.False_)

In [10]:
df["Year"] = df["Date"].dt.year

In [11]:
df["Month"] = df["Date"].dt.month

In [12]:
years = df["Year"]
years.unique()

array([2020, 2021, 2022, 2023, 2024], dtype=int32)

In [13]:
months = df["Month"]
months.unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int32)

In [14]:
df["Month-Year"] = df["Date"].dt.to_period('M')

In [15]:
mounts_years = df["Month-Year"]
mounts_years.unique()

<PeriodArray>
['2020-01', '2020-02', '2020-03', '2020-04', '2020-05', '2020-06', '2020-07',
 '2020-08', '2020-09', '2020-10', '2020-11', '2020-12', '2021-01', '2021-02',
 '2021-03', '2021-04', '2021-05', '2021-06', '2021-07', '2021-08', '2021-09',
 '2021-10', '2021-11', '2021-12', '2022-01', '2022-02', '2022-03', '2022-04',
 '2022-05', '2022-06', '2022-07', '2022-08', '2022-09', '2022-10', '2022-11',
 '2022-12', '2023-01', '2023-02', '2023-03', '2023-04', '2023-05', '2023-06',
 '2023-07', '2023-08', '2023-09', '2023-10', '2023-11', '2023-12', '2024-01',
 '2024-02', '2024-03', '2024-04', '2024-05', '2024-06', '2024-07', '2024-08',
 '2024-09', '2024-10', '2024-11', '2024-12']
Length: 60, dtype: period[M]

## Questions
What is the time span of the dataset? - 2020-01 to 2024-12

Are there any suspicious dates? - No

Is the dataset complete across months? - Yes

In [16]:
# 4. Data Validation Rules:
# Check the types
types_of_expenses = df["Type"]
types_of_expenses.unique()

array(['Expense', 'Income'], dtype=object)

In [17]:
# Check the values (must be only positive)
(df["Amount"] <= 0).any()

np.False_

Check if the types have relatively valid data

In [18]:
group_expenses = df[df["Type"] == "Expense"]["Amount"]
group_expenses.mean(), group_expenses.std(), group_expenses.count(), group_expenses.max(), group_expenses.min()

(np.float64(1004.2507119476269),
 571.5369855807375,
 np.int64(1222),
 1999.82,
 14.37)

Expenses mean is 1004, std is 571, so 2std = 2146, that means max number 1999 is not outliner same with min -2std=-138

In [19]:
group_incomes = df[df["Type"] == "Income"]["Amount"]
group_incomes.mean(), group_incomes.std(), group_incomes.count(), group_incomes.max(), group_incomes.min()

(np.float64(2640.6007194244603),
 1261.8904142661556,
 np.int64(278),
 4996.0,
 500.0)

Income mean is 2640, std is 1261, so 2std = 5162, that means max number 4996 is not outliner same with min -2std=-118

In [34]:
# What is the most and min amount of transactions in a day?
daily_counts = df.groupby('Date').size()
daily_counts.min(), daily_counts.max()

(1, 6)

In [40]:
category_count = df.groupby("Category")["Amount"].sum()
category_combined = df.groupby("Category").size()
category_count, category_combined

(Category
 Entertainment       148165.47
 Food & Drink        159493.39
 Health & Fitness    145194.06
 Investment          363252.00
 Other               370835.00
 Rent                162075.39
 Salary              149053.55
 Shopping            146880.75
 Travel              169497.79
 Utilities           146833.97
 Name: Amount, dtype: float64,
 Category
 Entertainment       143
 Food & Drink        149
 Health & Fitness    152
 Investment          142
 Other               136
 Rent                165
 Salary              146
 Shopping            150
 Travel              160
 Utilities           157
 dtype: int64)

5. Deliverable

1) Short brief summary

. Dataset contains both Income and Expense transactions.

. No negative or zero Amount values detected.

. No global statistical outliers using 2 standard deviation rule.

. Daily transaction counts range from 1 to 6 (realistic behavior).

. No immediate impossible transaction frequency patterns detected.

2) Cleaning Actions Performed

.Converted Date column to datetime format.

.Verified sign consistency between Type and Amount.

.Validated transaction frequency per day.

.Checked for statistical outliers globally.

3) Assumptions Made
 
.Dataset likely represents a mixed income profile (investment + salary + other).

.Income and Expense values are stored as positive numbers; direction is defined by Type column.

.Large income values (up to ~5000) are considered valid based on statistical boundaries.

.No abnormal behavioral spikes detected at daily level.

PHASE 2 — SPENDING STRUCTURE ANALYSIS